### PR TITLE
Improve Iceberg Glue table caching

### PR DESCRIPTION
**Pull request by alexjo2144**
## Description

Reduce the number of getTable calls by caching metadata any time a table is retrieved from Glue.

> Is this change a fix, improvement, new feature, refactoring, or other?

Improvement

> Is this a change to the core query engine, a connector, client library, or the SPI interfaces? (be specific)

Iceberg connector with Glue

> How would you describe this change to a non-technical end user or system administrator?

Reduce the number of repeated API calls

## Documentation

( ) No documentation is needed.
( ) Sufficient documentation is included in this PR.
( ) Documentation PR is available with #prnumber.
( ) Documentation issue #issuenumber is filed, and can be handled later.

## Release notes

( ) No release notes entries required.
(x) Release notes entries required with the following suggested text:

```markdown
# Iceberg
* Reduce the number of redundant API calls to AWS Glue.
```

### CODE REVIEW

**Comment by findinpath** [comment]
While skimming through `TrinoGlueCatalog` I've noticed that the cache holds the entries indefinitely 

https://github.com/trinodb/trino/blob/6310e5e1415fe0bc89444016fc516640c3a3fcbb/plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/glue/TrinoGlueCatalog.java#L119


While holding an entry in the Trino cache, could the drop of the corresponding table via spark and its recreation again via spark  lead to an unexpected situation in Trino?

**Comment by alexjo2144** [comment]
> While skimming through TrinoGlueCatalog I've noticed that the cache holds the entries indefinitely

Yes, but each transaction (query because we don't really support transactions) gets a fresh instance of the `TrinoGlueCatalog` so it's not a long lived cache.

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/glue/TrinoGlueCatalog.java [hunk 1]
```diff
@@ -565,7 +588,11 @@ public List<SchemaTableName> listViews(ConnectorSession session, Optional<String
     @Override
     public Optional<ConnectorViewDefinition> getView(ConnectorSession session, SchemaTableName viewName)
     {
-        Optional<com.amazonaws.services.glue.model.Table> table = getTable(viewName);
+        if (tableMetadataCache.containsKey(viewName)) {
+            return Optional.empty();
```
Comment (findepi) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/glue/TrinoGlueCatalog.java:592]: Add explanatory comment

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/glue/TrinoGlueCatalog.java [hunk 2]
```diff
@@ -724,7 +751,11 @@ private void dropStorageTable(ConnectorSession session, com.amazonaws.services.g
     @Override
     protected Optional<ConnectorMaterializedViewDefinition> doGetMaterializedView(ConnectorSession session, SchemaTableName viewName)
     {
-        Optional<com.amazonaws.services.glue.model.Table> maybeTable = getTable(viewName);
+        if (tableMetadataCache.containsKey(viewName)) {
+            return Optional.empty();
```
Comment (findepi) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/glue/TrinoGlueCatalog.java:755]: Add a comment

File: plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/catalog/glue/TestIcebergGlueCatalogAccessOperations.java [hunk 3]
```diff
@@ -160,7 +160,7 @@ public void testSelect()
 
             assertGlueMetastoreApiInvocations("SELECT * FROM test_select_from",
                     ImmutableMultiset.builder()
-                            .addCopies(GET_TABLE, 3)
+                            .add(GET_TABLE)
```
Comment (findepi) [inline plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/catalog/glue/TestIcebergGlueCatalogAccessOperations.java:163]: so nice!

File: plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/catalog/glue/TestIcebergGlueCatalogAccessOperations.java [hunk 4]
```diff
@@ -211,7 +211,7 @@ public void testSelectFromViewWithFilter()
 
             assertGlueMetastoreApiInvocations("SELECT * FROM test_select_view_where_view WHERE age = 2",
                     ImmutableMultiset.builder()
-                            .addCopies(GET_TABLE, 5)
+                            .addCopies(GET_TABLE, 3)
```
Comment (findepi) [inline plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/catalog/glue/TestIcebergGlueCatalogAccessOperations.java:214]: 2 more to go :)

**Review by findepi — APPROVED** [review]

**Comment by alexjo2144** [comment]
Added comments, thanks for the review @findepi

**Comment by alexjo2144** [comment]
@findepi can you take another look when you get a chance? I added a fixup that adds the same approach for views and MVs

**Review by findinpath — APPROVED** [review]

**Review by ebyhr — APPROVED** [review]

**Comment by ebyhr** [comment]
Could you confirm CI failure? 
```
Error:  io.trino.plugin.iceberg.catalog.glue.TestTrinoGlueCatalogTest.testView  Time elapsed: 2.059 s  <<< FAILURE!
java.lang.AssertionError: 

Expecting an empty Optional but was containing value: <ConnectorViewDefinition[owner=user, runAsInvoker=false, columns=[name type:[varchar(25)]], originalSql=[SELECT name FROM local.tiny.nation]]>.
	at io.trino.plugin.iceberg.BaseTrinoCatalogTest.testView(BaseTrinoCatalogTest.java:235)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:104)
	at org.testng.internal.Invoker.invokeMethod(Invoker.java:645)
	at org.testng.internal.Invoker.invokeTestMethod(Invoker.java:851)
	at org.testng.internal.Invoker.invokeTestMethods(Invoker.java:1177)
	at org.testng.internal.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:129)
	at org.testng.internal.TestMethodWorker.run(TestMethodWorker.java:112)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.lang.Thread.run(Thread.java:833)
```

**Comment by findinpath** [comment]
@alexjo2144  we should probably drop the cached entry when renaming/dropping tables/views/MVs.

**Comment by alexjo2144** [comment]
I'll add it in for the test, but the way we're set up now that should never happen. Since a new catalog is created per query, the use case of 
- `catalog.getView()`
- `catalog.dropView()`
- `catalog.getView()`

shouldn't happen

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/glue/TrinoGlueCatalog.java [hunk 5]
```diff
@@ -117,6 +120,8 @@
     private final GlueMetastoreStats stats;
 
     private final Map<SchemaTableName, TableMetadata> tableMetadataCache = new ConcurrentHashMap<>();
+    private final Map<SchemaTableName, ConnectorViewDefinition> viewCache = new ConcurrentHashMap<>();
+    private final Map<SchemaTableName, ConnectorMaterializedViewDefinition> materializedViewCache = new ConcurrentHashMap<>();
```
Comment (findepi) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/glue/TrinoGlueCatalog.java:124]: this & tableMetadataCache -- they should never be out of sync.
maybe we should have a single cache with `T | V | MV` as the value?

File: plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/catalog/glue/TestIcebergGlueCatalogAccessOperations.java [hunk 6]
```diff
@@ -265,7 +266,8 @@ public void testRefreshMaterializedView()
 
             assertGlueMetastoreApiInvocations("REFRESH MATERIALIZED VIEW test_refresh_mview_view",
                     ImmutableMultiset.builder()
-                            .addCopies(GET_TABLE, 5)
+                            .addCopies(GET_TABLE, 9)
+                            .addCopies(UPDATE_TABLE, 2)
```
Comment (findepi) [inline plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/catalog/glue/TestIcebergGlueCatalogAccessOperations.java:270]: 6 more calls. Why?

File: plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/catalog/glue/TestIcebergGlueCatalogAccessOperations.java [hunk 7]
```diff
@@ -365,37 +367,37 @@ public void testSelectSystemTable()
...
                     ImmutableMultiset.builder()
-                            .addCopies(GET_TABLE, 1)
+                            .add(GET_TABLE)
                             .build());
 
             // select from $partitions
             assertGlueMetastoreApiInvocations("SELECT * FROM \"test_select_snapshots$partitions\"",
                     ImmutableMultiset.builder()
-                            .addCopies(GET_TABLE, 1)
+                            .add(GET_TABLE)
```
Comment (findepi) [inline plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/catalog/glue/TestIcebergGlueCatalogAccessOperations.java:388]: This & others

Unnecessary change.

(i don't want to argue whether `.add(x)` vs `.add(x, 1)` should be used, but let's treat it as an unrelated change)

File: plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/catalog/glue/TestIcebergGlueCatalogAccessOperations.java (hunk 6 above)
Comment (alexjo2144) [reply plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/catalog/glue/TestIcebergGlueCatalogAccessOperations.java:270]: The test was disabled before, so the value here was not valid

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/glue/TrinoGlueCatalog.java (hunk 5 above)
Comment (alexjo2144) [reply plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/glue/TrinoGlueCatalog.java:124]: I can try that, but what do you mean by "out of sync"?

File: plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/catalog/glue/TestIcebergGlueCatalogAccessOperations.java (hunk 6 above)
Comment (findepi) [reply plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/catalog/glue/TestIcebergGlueCatalogAccessOperations.java:270]: I missed that. Can you make enabling tests a separate commit?

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/glue/TrinoGlueCatalog.java (hunk 5 above)
Comment (findepi) [reply plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/glue/TrinoGlueCatalog.java:124]: Technically, it's possible that two or three caches contain an entry with same key.
This would lead to inconsistent state (eg we would report given name to be a table, and also a view), since order of checking these caches is different in respective methods.

Comment (findepi) [reply plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/glue/TrinoGlueCatalog.java:124]: > a single cache with T | V | MV as the value

i don't know whether this would be pretty or ugly.
maybe fixing the ordering to be the same would be enough

Comment (alexjo2144) [reply plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/glue/TrinoGlueCatalog.java:123]: I added some more verification, but I don't think this can happen. Anywhere the caches are set has the pattern:
- Check if the relevant cache has the key, if so return
- Check if any of the other caches has an entry, if so return
- Get and cache the entry

**Comment by findepi** [comment]
@alexjo2144 please see the CI

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/glue/TrinoGlueCatalog.java (hunk 5 above)
Comment (findepi) [reply plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/glue/TrinoGlueCatalog.java:126]: i think it's fine

**Comment by findepi** [comment]
(squashed)

**Comment by findepi** [comment]
(squashed)

**Comment by findepi** [comment]
Updated https://github.com/trinodb/trino/pull/13950 too

**Comment by alexjo2144** [comment]
CI hit the 60 minute timeout, I put this PR up to split the cloud tests into a separate CI job https://github.com/trinodb/trino/pull/14081

**Comment by alexjo2144** [comment]
@findepi @ebyhr would one of you mind kicking off another CI with secrets?

**Comment by findepi** [comment]
@alexjo2144 done

**Comment by alexjo2144** [comment]
Thanks, will take a look at failures tomorrow

**Comment by alexjo2144** [comment]
I had the wrong type of exception. Updated in the fixup commit

**Comment by findinpath** [comment]
@findepi could you please run the PR again with secrets?

File: plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/catalog/glue/TestIcebergGlueCatalogAccessOperations.java [hunk 8]
```diff
@@ -347,7 +350,7 @@ public void testShowStatsForTableWithFilter()
 
             assertGlueMetastoreApiInvocations("SHOW STATS FOR (SELECT * FROM test_show_stats_with_filter where age >= 2)",
                     ImmutableMultiset.builder()
-                            .addCopies(GET_TABLE, 3)
+                            .addCopies(GET_TABLE, 1)
```
Comment (findinpath) [inline plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/catalog/glue/TestIcebergGlueCatalogAccessOperations.java:353]: nit: just to be in sync with the other changes from your commit you could use here:

```
.add(GET_TABLE)
```

**Comment by findepi** [comment]
Updated https://github.com/trinodb/trino/pull/13950

**Comment by alexjo2144** [comment]
> Error:  io.trino.plugin.iceberg.TestIcebergGcsConnectorSmokeTest.testDeleteRowsConcurrently

Failures are unrelated. The variability on the GCS tests is really high though. The last run on master needed about 30 seconds for that test on GCS which is a long time, but still a lot less than the 2 minute timeout. 

Possibly related to: https://github.com/trinodb/trino/pull/3507

**Comment by alexjo2144** [comment]
Failures are just in the Cassandra connector tests
